addon/-private/system/store.js

```diff
@@ -2566,15 +2566,23 @@ Store = Service.extend({
 
     assert(`You can no longer pass a modelClass as the first argument to store._buildInternalModel. Pass modelName instead.`, typeof modelName === 'string');
 
-    let recordMap = this._internalModelsFor(modelName);
+    let internalModels = this._internalModelsFor(modelName);
+    let existingInternalModel = internalModels.get(id);
 
-    assert(`The id ${id} has already been used with another record for modelClass '${modelName}'.`, !id || !recordMap.get(id));
+    if (existingInternalModel) {
```

Comment:
Mentioned this in chat, but I think this should be `if (existingInternalModel && existingInternalModel._isDematerializing) {` (otherwise we destroy everything and the assert just below is useless).